### PR TITLE
Search

### PR DESCRIPTION
**Pull request by pavolfranekdev**

### CODE REVIEW

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 1]
```diff
@@ -65,7 +78,147 @@ class MainActivity : SimpleActivity() {
...
+                currentTextFragment?.setTextWatcher()
+            }
+        }
+
+        search_previous.setOnClickListener {
+            currentNotesView()?.let { noteView ->
+                if (searchIndex > 0) {
+                    searchIndex--
+                }
+                else {
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: move such elses a line up, to make it "} else {". On other places too.

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 2]
```diff
@@ -65,7 +78,147 @@ class MainActivity : SimpleActivity() {
...
+                    searchIndex = 0
+                }
+
+                selectMatch(noteView)
+            }
+        }
+
+        search_clear.setOnClickListener {
+            if (search_query.value.isNotBlank())
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: missing brackets

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 3]
```diff
@@ -65,7 +78,147 @@ class MainActivity : SimpleActivity() {
...
+
+    private fun searchMatches(textToHighlight: String, content: String): ArrayList<Int> {
+        val indexes = arrayListOf<Int>()
+        var indexOf = content.indexOf(textToHighlight, 0, ignoreCase = true)
+
+        var offset = 0
+        while (offset < content.length && indexOf != -1) {
+            indexOf = content.indexOf(textToHighlight, offset, ignoreCase = true)
+
+            if (indexOf == -1)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: missing brackets

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/extensions/String.kt [hunk 4]
```diff
@@ -14,3 +19,7 @@ fun String.parseChecklistItems(): ArrayList<ChecklistItem>? {
     }
     return null
 }
+
+@RequiresApi(Build.VERSION_CODES.N)
+fun String.toHtml() = if (isNougatPlus()) Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY) else Html.fromHtml(this)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/extensions/String.kt]: split it into multiple lines

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 5]
```diff
@@ -114,6 +267,7 @@ class MainActivity : SimpleActivity() {
             findItem(R.id.open_note).isVisible = shouldBeVisible
             findItem(R.id.delete_note).isVisible = shouldBeVisible
             findItem(R.id.export_all_notes).isVisible = shouldBeVisible
+            findItem(R.id.open_search).isVisible = view_pager.currentItem != 0
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: why isnt search shown at the default page?

**Comment by tibbi** [comment]
some new notes:
1. lets close Search immediately after pressing X for the first time, dont just clear the input
2. once a search string is found and it isnt visible on the screen, jump to it automatically like if the user pressed the Right chevron once
3. highlight the current string occurrence that you are switching between with the arrows. Maybe reduce the alpha of the accent color of all ocurrences, keep only the current ocurrence on max.

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt (hunk 5 above)
Comment (pavolfranekdev) [reply app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: fixed

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 6]
```diff
@@ -191,9 +199,10 @@ class MainActivity : SimpleActivity() {
         while (offset < content.length && indexOf != -1) {
             indexOf = content.indexOf(textToHighlight, offset, true)
 
-            if (indexOf == -1) break
-            else {
-                wordToSpan.setSpan(BackgroundColorSpan(color(R.color.color_accent)), indexOf, indexOf + textToHighlight.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
+            if (indexOf == -1) {
+                break
+            } else {
+                wordToSpan.setSpan(BackgroundColorSpan(ColorUtils.setAlphaComponent(config.primaryColor, 90)), indexOf, indexOf + textToHighlight.length, Spannable.SPAN_EXCLUSIVE_EXCLUSIVE)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: make this line shorter, split it into some variables

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 7]
```diff
@@ -235,6 +244,11 @@ class MainActivity : SimpleActivity() {
             setTextColor(config.textColor)
         }
         updateTextColors(view_pager)
+
+        search_root.setBackgroundColor(config.primaryColor)
+        search_previous.applyColorFilter(config.primaryColor.getContrastColor())
+        search_next.applyColorFilter(config.primaryColor.getContrastColor())
+        search_clear.applyColorFilter(config.primaryColor.getContrastColor())
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: store contrastColor in some variable above, do not fetch it 3x

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 8]
```diff
@@ -342,6 +356,7 @@ class MainActivity : SimpleActivity() {
         view_pager.currentItem = getWantedNoteIndex(wantedNoteId)
         checkIntents(intent)
     }
+    private val currentItemIsCheckList get() = mAdapter?.isChecklistFragment(view_pager.currentItem) ?: false
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: add an empty line above it

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/extensions/String.kt [hunk 9]
```diff
@@ -21,5 +21,9 @@ fun String.parseChecklistItems(): ArrayList<ChecklistItem>? {
 }
 
 @RequiresApi(Build.VERSION_CODES.N)
-fun String.toHtml() = if (isNougatPlus()) Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY) else Html.fromHtml(this)
+fun String.toHtml() =
+        if (isNougatPlus())
+            Html.fromHtml(this, Html.FROM_HTML_MODE_LEGACY)
+        else
+            Html.fromHtml(this)
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/extensions/String.kt]: missing brackets

File: app/src/main/res/layout/search_item.xml [hunk 10]
```diff
@@ -1,53 +1,68 @@
...
+            android:padding="@dimen/small_margin"
+            android:src="@drawable/ic_cross_vector" />
+    </LinearLayout>
+
+    <View
+        android:layout_width="match_parent"
+        android:layout_height="1dp"
+        android:layout_gravity="bottom"
+        android:alpha="0.1"
+        android:background="#fff" />
```
Comment (tibbi) [inline app/src/main/res/layout/search_item.xml]: always use 6-8 characters at hex codes, but in this case you can also use color/md_grey_white, it is in commons

File: app/src/main/res/layout/search_item.xml [hunk 11]
```diff
@@ -0,0 +1,68 @@
...
+    android:gravity="center_vertical"
+    android:orientation="vertical"
+    android:visibility="gone"
+    tools:visibility="visible">
+
+    <LinearLayout
+        android:layout_width="match_parent"
+        android:layout_height="wrap_content"
+        android:orientation="horizontal"
+        android:paddingTop="@dimen/medium_margin"
```
Comment (tibbi) [inline app/src/main/res/layout/search_item.xml]: add android:gravity="center_vertical" for proper centering

**Comment by tibbi** [comment]
lets try adding 2 more things for better UX.
1. while the focus is on the search field, make "Enter" behave like pressing the "Next" chevron to make jumping between ocurrences easier
2. while the search is open, make pressing Back just close it, not exit the whole app

File: app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt [hunk 12]
```diff
@@ -65,7 +79,157 @@ class MainActivity : SimpleActivity() {
...
+        val wordToSpan = SpannableString(view.text)
+
+        var offset = 0
+        while (offset < content.length && indexOf != -1) {
+            indexOf = content.indexOf(highlightText, offset, true)
+
+            if (indexOf == -1) {
+                break
+            } else {
+                val spanBgColor = BackgroundColorSpan(ColorUtils.setAlphaComponent(config.primaryColor, 90))
```
Comment (tibbi) [inline app/src/main/kotlin/com/simplemobiletools/notes/pro/activities/MainActivity.kt]: lets make this 128 instead of 90 to make it easier to see

**Pull request merged**